File: lib/trades.ts
import {
  classifyTransaction,
  fetchAllCongressTrades,
  fetchMemberCongressTrades,
  findMatchingPurchase,
  findMatchingSale,
  parseTradeRange,
  type RawCongressTrade,
} from "./quiver"
import { getDaySnapshot, getLatestSnapshot, type DaySnapshot } from "./prices"

export type TradeLeg = { date: string; range: string; transactionType: string }

export type ProfitLoss = {
  buyPrice: number
  exitPrice: number
  exitBasis: "sale" | "current"
  pctChange: number
  plLow: number
  plHigh: number
}

export type TradeDetail = {
  id: string
  ticker: string
  assetName: string
  memberName: string
  bioguideId: string
  chamber: string
  party: string
  buy: TradeLeg | null
  sell: TradeLeg | null
  buySnapshot: DaySnapshot | null
  sellSnapshot: DaySnapshot | null
  todaySnapshot: DaySnapshot | null
  profitLoss: ProfitLoss | null
}

function toLeg(trade: RawCongressTrade): TradeLeg {
  return {
    date: trade.Date ?? "",
    range: trade.Range ?? "",
    transactionType: trade.Transaction ?? "",
  }
}

function computeProfitLoss(
  buyPrice: number | null,
  sellSnapshot: DaySnapshot | null,
  todaySnapshot: DaySnapshot | null,
  range: { low: number; high: number } | null
): ProfitLoss | null {
  if (buyPrice == null || buyPrice <= 0 || !range) return null

  let exitPrice: number | null = null
  let exitBasis: ProfitLoss["exitBasis"] = "current"
  if (sellSnapshot?.close != null) {
    exitPrice = sellSnapshot.close
    exitBasis = "sale"
  } else if (todaySnapshot?.close != null) {
    exitPrice = todaySnapshot.close
    exitBasis = "current"
  }
  if (exitPrice == null) return null

  const pctChange = ((exitPrice - buyPrice) / buyPrice) * 100
  return {
    buyPrice,
    exitPrice,
    exitBasis,
    pctChange,
    plLow: range.low * (pctChange / 100),
    plHigh: range.high * (pctChange / 100),
  }
}

export async function loadTradeDetail(id: string): Promise<TradeDetail | null> {
  const apiKey = process.env.QUIVER_API_KEY
  if (!apiKey) return null
  const bioguide = id.split("|")[0] ?? ""
  let all: RawCongressTrade[]
  try {
    all = bioguide ? await fetchMemberCongressTrades(bioguide, apiKey) : []
    if (!all.some((t) => String(t.UniqueID) === id)) {
      all = await fetchAllCongressTrades(apiKey)
    }
  } catch {
    return null
  }

  const clicked = all.find((t) => String(t.UniqueID) === id)
  if (!clicked) return null
  const kind = classifyTransaction(clicked.Transaction)
  const buyTrade = kind === "sell" ? findMatchingPurchase(all, clicked) : clicked
  const sellTrade = kind === "sell" ? clicked : findMatchingSale(all, clicked)

  const ticker = clicked.Ticker ?? ""
  const hasTicker = Boolean(ticker) && ticker !== "-"

  const [buySnapshot, sellSnapshot, todaySnapshot] = await Promise.all([
    buyTrade?.Date && hasTicker
      ? getDaySnapshot(ticker, buyTrade.Date)
      : Promise.resolve(null),
    sellTrade?.Date && hasTicker
      ? getDaySnapshot(ticker, sellTrade.Date)
      : Promise.resolve(null),
    hasTicker ? getLatestSnapshot(ticker) : Promise.resolve(null),
  ])

  const profitLoss = computeProfitLoss(
    buySnapshot?.close ?? null,
    sellSnapshot,
    todaySnapshot,
    parseTradeRange(buyTrade?.Range)
  )

  return {
    id,
    ticker,
    assetName: clicked.AssetDescription ?? "",
    memberName: clicked.Representative ?? "",
    bioguideId: clicked.Bioguide ?? "",
    chamber: clicked.Chamber ?? "",
    party: clicked.Party ?? "",
    buy: buyTrade ? toLeg(buyTrade) : null,
    sell: sellTrade ? toLeg(sellTrade) : null,
    buySnapshot,
    sellSnapshot,
    todaySnapshot,
    profitLoss,
  }
}

File: backend/app/services/trades.py
"""Single-trade detail with price snapshots and P/L (port of lib/trades.ts)."""

import asyncio

from ..clients.prices import get_day_snapshot, get_latest_snapshot
from ..clients.quiver import (
    classify_transaction,
    fetch_all_congress_trades,
    fetch_member_congress_trades,
    find_matching_purchase,
    find_matching_sale,
    parse_trade_range,
)
from ..config import quiver_api_key


def _to_leg(trade: dict) -> dict:
    return {
        "date": trade.get("Date") or "",
        "range": trade.get("Range") or "",
        "transactionType": trade.get("Transaction") or "",
    }


def _compute_profit_loss(
    buy_price: float | None,
    sell_snapshot: dict | None,
    today_snapshot: dict | None,
    trade_range: dict | None,
) -> dict | None:
    if buy_price is None or buy_price <= 0 or not trade_range:
        return None

    exit_price: float | None = None
    exit_basis = "current"
    if sell_snapshot and sell_snapshot.get("close") is not None:
        exit_price = sell_snapshot["close"]
        exit_basis = "sale"
    elif today_snapshot and today_snapshot.get("close") is not None:
        exit_price = today_snapshot["close"]
        exit_basis = "current"
    if exit_price is None:
        return None

    pct_change = ((exit_price - buy_price) / buy_price) * 100
    return {
        "buyPrice": buy_price,
        "exitPrice": exit_price,
        "exitBasis": exit_basis,
        "pctChange": pct_change,
        "plLow": trade_range["low"] * (pct_change / 100),
        "plHigh": trade_range["high"] * (pct_change / 100),
    }


async def load_trade_detail(trade_id: str) -> dict | None:
    api_key = quiver_api_key()
    if not api_key:
        return None
    bioguide = trade_id.split("|")[0] if trade_id else ""
    try:
        all_trades = (
            await fetch_member_congress_trades(bioguide, api_key) if bioguide else []
        )
        if not any(str(t.get("UniqueID")) == trade_id for t in all_trades):
            all_trades = await fetch_all_congress_trades(api_key)
    except Exception:
        return None

    clicked = next(
        (t for t in all_trades if str(t.get("UniqueID")) == trade_id), None
    )
    if not clicked:
        return None
    kind = classify_transaction(clicked.get("Transaction"))
    buy_trade = find_matching_purchase(all_trades, clicked) if kind == "sell" else clicked
    sell_trade = clicked if kind == "sell" else find_matching_sale(all_trades, clicked)

    ticker = clicked.get("Ticker") or ""
    has_ticker = bool(ticker) and ticker != "-"

    async def none() -> None:
        return None

    buy_snapshot, sell_snapshot, today_snapshot = await asyncio.gather(
        get_day_snapshot(ticker, buy_trade["Date"])
        if buy_trade and buy_trade.get("Date") and has_ticker
        else none(),
        get_day_snapshot(ticker, sell_trade["Date"])
        if sell_trade and sell_trade.get("Date") and has_ticker
        else none(),
        get_latest_snapshot(ticker) if has_ticker else none(),
    )

    profit_loss = _compute_profit_loss(
        (buy_snapshot or {}).get("close"),
        sell_snapshot,
        today_snapshot,
        parse_trade_range(buy_trade.get("Range") if buy_trade else None),
    )

    return {
        "id": trade_id,
        "ticker": ticker,
        "assetName": clicked.get("AssetDescription") or "",
        "memberName": clicked.get("Representative") or "",
        "bioguideId": clicked.get("Bioguide") or "",
        "chamber": clicked.get("Chamber") or "",
        "party": clicked.get("Party") or "",
        "buy": _to_leg(buy_trade) if buy_trade else None,
        "sell": _to_leg(sell_trade) if sell_trade else None,
        "buySnapshot": buy_snapshot,
        "sellSnapshot": sell_snapshot,
        "todaySnapshot": today_snapshot,
        "profitLoss": profit_loss,
    }
